bridges/modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -62,153 +95,276 @@ const MESSAGE_SIZE_FEE_BASE: FixedU128 = FixedU128::from_rational(1, 1000); // 0
+	pub type BridgeIdOf<T, I> = <<T as Config<I>>::BridgeIdResolver as ResolveBridgeId>::BridgeId;
+
 	#[pallet::pallet]
 	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
 
 	#[pallet::hooks]
 	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
-		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
-			// if XCM channel is still congested, we don't change anything
-			if T::LocalXcmChannelManager::is_congested(&T::SiblingBridgeHubLocation::get()) {
-				return T::WeightInfo::on_initialize_when_congested()
+		fn on_idle(_n: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			let mut meter = WeightMeter::with_limit(remaining_weight);
+
+			// Iterate all congested bridges
+			let mut bridges_to_update = Vec::new();
+			let mut bridges_to_remove = Vec::new();
+			for (bridge_id, mut bridge_state) in Bridges::<T, I>::iter() {
+				meter.consume(T::DbWeight::get().reads(1));
+
+				// If no longer congested, we can start decreasing the fee factor.
+				if !bridge_state.is_congested {
+					let previous_factor = bridge_state.delivery_fee_factor;
+					let new_factor = previous_factor / EXPONENTIAL_FEE_BASE;
+					if new_factor >= MINIMAL_DELIVERY_FEE_FACTOR {
+						bridge_state.delivery_fee_factor = new_factor;
+						if meter
+							.try_consume(T::WeightInfo::on_idle_when_bridge_state_updated())
+							.is_err()
+						{
+							break;
+						}
+						bridges_to_update.push((bridge_id, previous_factor, bridge_state));
+					} else {
+						if meter
+							.try_consume(T::WeightInfo::on_idle_when_bridge_state_removed())
+							.is_err()
+						{
+							break;
+						}
+						bridges_to_remove.push((bridge_id, previous_factor));
+					}
+				}
 			}
 
-			// if we can't decrease the delivery fee factor anymore, we don't change anything
-			let mut delivery_fee_factor = Self::delivery_fee_factor();
-			if delivery_fee_factor == MINIMAL_DELIVERY_FEE_FACTOR {
-				return T::WeightInfo::on_initialize_when_congested()
+			// remove
+			for (bridge_id, previous_value) in bridges_to_remove.into_iter() {
+				log::info!(
+					target: LOG_TARGET,
+					"Bridge channel with id {:?} is uncongested. Removing fee factor!",
+					bridge_id,
+				);
+				Bridges::<T, I>::remove(&bridge_id);
+				Self::deposit_event(Event::DeliveryFeeFactorDecreased {
+					previous_value,
+					new_value: 0.into(),
+					bridge_id,
+				});
+			}
+			// update
+			for (bridge_id, previous_value, bridge_state) in bridges_to_update.into_iter() {
+				let new_value = bridge_state.delivery_fee_factor;
+				log::info!(
+					target: LOG_TARGET,
+					"Bridge channel with id {:?} is uncongested. Decreasing fee factor from {} to {}!",
+					bridge_id,
+					previous_value,
+					new_value,
+				);
+				Bridges::<T, I>::insert(&bridge_id, bridge_state);
+				Self::deposit_event(Event::DeliveryFeeFactorDecreased {
+					previous_value,
+					new_value,
+					bridge_id,
+				});
 			}
```

Comment:
Just curious why not guard the weight usage with `meter` in second loop, what if there is a lot of bridges to update/remove here while the weight is used up in first loop?